### PR TITLE
Janosch/new client

### PR DESCRIPTION
**Pull request by jgraef**
## Pull request checklist

- [X] All tests pass. Demo project builds and runs.
- [X] I have resolved any merge conflicts.

## What's in this pull request?

This improves the new client API `nimiq_lib2` and implements a new `nimiq_client2` on top of it. There's still a ton of features missing:

 * Refactor extras: contains logging, deadlock detection, panic handling.
 * Even easier launcher, that allows constructing a client with minimal effort, e.g.`nimiq.launch()`. This would take care of loading config and initializing extras, but should also allow customization.
 * Integration of RPC and metrics server.
 * Loading config from environment variables.

Once the RPC and metrics server are integrated we can switch this out with `nimiq_lib` and `nimiq_client`.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -32,6 +32,7 @@ members = [
     "consensus",
     "network",
     "client",
+    "client2",
```
Comment (paberr) [inline Cargo.toml:35]: When do we get rid of `client` and rename `client2` to `client`? 😆

File: blockchain-albatross/src/blockchain.rs [hunk 2]
```diff
@@ -92,7 +92,7 @@ pub struct Blockchain {
     pub network_id: NetworkId,
     // TODO network_time: Arc<NetworkTime>,
     pub notifier: RwLock<Notifier<'static, BlockchainEvent>>,
-    pub(crate) chain_store: Arc<ChainStore>,
+    pub chain_store: Arc<ChainStore>,
```
Comment (paberr) [inline blockchain-albatross/src/blockchain.rs:95]: Theoretically allows changing the database, which is not ideal. But then we also have an Arc to the database.

File: blockchain-albatross/src/blockchain.rs [hunk 3]
```diff
@@ -101,9 +101,9 @@ pub struct Blockchain {
 }
 
 pub struct BlockchainState {
-    accounts: Accounts,
-    transaction_cache: TransactionCache,
-    pub(crate) reward_registry: SlashRegistry,
+    pub accounts: Accounts,
```
Comment (paberr) [inline blockchain-albatross/src/blockchain.rs:104]: Also allows changing the Accounts state.

File: network/src/connection/connection_pool.rs [hunk 4]
```diff
@@ -444,7 +444,10 @@ impl<B: AbstractBlockchain + 'static> ConnectionPool<B> {
 
     /// Initiates a outbound connection.
     pub fn connect_outbound(&self, peer_address: Arc<PeerAddress>) -> bool {
+        trace!("connect_outbound({})", peer_address);
```
Comment (paberr) [inline network/src/connection/connection_pool.rs]: To be removed.

File: network/src/connection/connection_pool.rs [hunk 5]
```diff
@@ -444,7 +444,10 @@ impl<B: AbstractBlockchain + 'static> ConnectionPool<B> {
...
         let _guard = self.change_lock.lock();
+        trace!("acquired lock");
```
Comment (paberr) [inline network/src/connection/connection_pool.rs]: To be removed.

File: network/src/connection/connection_pool.rs [hunk 6]
```diff
@@ -461,9 +464,12 @@ impl<B: AbstractBlockchain + 'static> ConnectionPool<B> {
             },
         };
 
+        trace!("got connection handle");
```
Comment (paberr) [inline network/src/connection/connection_pool.rs]: To be removed.

File: network/src/connection/connection_pool.rs [hunk 7]
```diff
@@ -461,9 +464,12 @@ impl<B: AbstractBlockchain + 'static> ConnectionPool<B> {
...
+
         // Create fresh ConnectionInfo instance.
         let mut state = self.state.write();
         let connection_id = state.add(ConnectionInfo::outbound(peer_address.clone()));
+        trace!("connection_id = {}", connection_id);
```
Comment (paberr) [inline network/src/connection/connection_pool.rs]: To be removed.

File: network/src/network.rs [hunk 8]
```diff
@@ -182,6 +182,8 @@ impl<B: AbstractBlockchain> Network<B> {
     }
 
     fn check_peer_count(&self) {
+        trace!("check_peer_count: seeded={}, good_peer_set={}", self.addresses.seeded(), self.scorer.read().is_good_peer_set());
```
Comment (paberr) [inline network/src/network.rs]: To be removed.

File: network/src/peer_channel/channel.rs [hunk 9]
```diff
@@ -40,6 +40,8 @@ pub struct PeerChannel {
 
 impl PeerChannel {
     pub fn new(network_connection: &NetworkConnection) -> Self {
+        trace!("new peer channel: {:?}", network_connection.net_address());
```
Comment (paberr) [inline network/src/peer_channel/channel.rs]: To be removed.

File: Cargo.toml (hunk 1 above)
Comment (jgraef) [reply Cargo.toml:35]: After `client2` supports RPC and metrics server. Feel free to implement that.

File: blockchain-albatross/src/blockchain.rs (hunk 3 above)
Comment (jgraef) [reply blockchain-albatross/src/blockchain.rs:104]: So, how am I supposed to read an account balance then?

File: blockchain-albatross/src/blockchain.rs (hunk 2 above)
Comment (jgraef) [reply blockchain-albatross/src/blockchain.rs:95]: What we could do is have all important objects wrapped into public facing types. This way we it's also easier to keep the public-facing API stable. We will still have fun with semantic versioning on all the other crates though.

**Review by paberr — APPROVED** [review]

**Pull request merged**